domains/client/domain-operator/src/operator.rs

```diff
@@ -248,89 +247,111 @@ where
-                    if block_number >= target_block_number.into() {
-                        break;
+                            if block_number >= target_block_number.into() {
+                                break;
+                            }
+                        }
+
+                        // Restore parameters
+                        unsafe {
+                            params.operator_streams.block_importing_notification_stream =
+                                Box::<IBNS>::into_inner(Pin::into_inner_unchecked(
+                                    block_importing_notification_stream,
+                                ));
+                            params.operator_streams.imported_block_notification_stream =
+                                Box::<CIBNS>::into_inner(Pin::into_inner_unchecked(
+                                    imported_block_notification_stream,
+                                ));
+                        }
                     }
-                }
 
-                // Restore parameters
-                unsafe {
-                    params.operator_streams.block_importing_notification_stream =
-                        Box::<IBNS>::into_inner(Pin::into_inner_unchecked(
-                            block_importing_notification_stream,
-                        ));
-                    params.operator_streams.imported_block_notification_stream =
-                        Box::<CIBNS>::into_inner(Pin::into_inner_unchecked(
-                            imported_block_notification_stream,
-                        ));
-                }
-            }
+                    crate::domain_worker::start_worker(
+                        spawn_essential.clone(),
+                        consensus_client.clone(),
+                        params.consensus_offchain_tx_pool_factory.clone(),
+                        params.maybe_operator_id,
+                        bundle_producer,
+                        bundle_processor.clone(),
+                        params.operator_streams,
+                    )
+                    .boxed()
+                };
+
+                spawn_essential.spawn_essential_blocking(
+                    "domain-operator-worker",
+                    None,
+                    Box::pin(start_worker_task),
+                );
 
-            crate::domain_worker::start_worker(
-                spawn_essential.clone(),
-                params.consensus_client.clone(),
-                params.consensus_offchain_tx_pool_factory.clone(),
-                params.maybe_operator_id,
-                bundle_producer,
-                bundle_processor.clone(),
-                params.operator_streams,
-            )
-            .boxed()
+                // Keep essential task running
+                pending::<()>().await;
+                Ok(())
+            }
         };
 
         spawn_essential.spawn_essential_blocking(
-            "domain-operator-worker",
+            "domain-operator-worker-starter",
```

Comment:
Sorry I'm unable to see why we need both `domain-operator-worker` and `domain-operator-worker-starter`, could give more detail about this?

Reply:
Not really deadlock, please try to make the whole `start_worker_task` an async block e.g. `let start_worker_task = async move { .. }` and removing `domain-operator-worker-starter` should be fine.

Reply:
Yeah, I understand what led to this change, my point is we don't need the `domain-operator-worker-starter` as a wrapper of `domain-operator-worker`. IIUC what we want is 3 sequential steps:
1. Get the `target_block_number`
2. Drain the steams
3. `start_worker`

This can be done by wrapping them into one async block:
```rust
let start_worker_task = async move {
	let target_block_number = target_block_receiver.recv().await;
	
	while let Some(...) = stream.next().await {
		...
	}

	crate::domain_worker::start_worker(...).boxed()
};
spawn_essential.spawn_essential_blocking(
	"domain-operator-worker",
	None,
	Box::pin(start_worker_task),
);
```

This doesn't work in the previous commit https://github.com/autonomys/subspace/pull/3115/commits/60d45e97d242c155c32ca4320b1567f11194ea5f because the `start_worker_task` is not an async block (i.e. `let start_worker_task = { ... }`) and is evaluated as a future of the last step`start_worker` but not a future of all the 3 steps, and first 2 steps will block the whole function `Operator::new()`